### DISCUSSION TITLE
Feedback on sketch of cache plugin

### DISCUSSION THREAD

**Discussion by forthrin** [HoneyryderChuck/httpx#50] [Q&A]
Made a quick sketch of a cache plugin. Would like feedback on:

1. If the new functionality are placed in the right modules/functions
2. How to pass/share options/variables between modules (most likely is wrong as sketched)

```ruby
module HTTPX::Plugins
  module Cache
    module InstanceMethods
      def initialize
        super
        @root ||= "#{Dir.home}/Downloads/Cache"
        @expire = 300
      end
    end

    module RequestMethods
      def request(method, url, **options)
        uri = URI.parse(url)
        @local = "#{@root}/#{uri.host}#{uri.request_uri}"
        if File.exist?(@local) && Time.now - File.mtime(@local) < (options[:expire] || @expire)
          @body = File.read(@local)
          puts "CACHE"
        else
          super
        end
      end
    end

    module ResponseMethods
      def body
        unless @body
          puts "GET"
          @body = super
          system("mkdir -p #{File.dirname(@local)}")
          File.write(@local, @body)
        end
        @body
      end
    end
  end
  register_plugin :cache, Cache
end

def test
  foo = HTTPX.plugin(:cache, root: '/tmp')
  url = 'http://localhost:4567/foo/bar/index.html'
  foo.get(url).body # GET
  foo.get(url, expire: 60).body # CACHE
  foo.get(url).html # CACHE + Nokogiri.HTML5()
end
```

**Comment by HoneyryderChuck** [comment]
Hi,

If it works for your use case, I'd say it's a 👍 . Other than that, do follow [the docs](https://gitlab.com/os85/httpx/-/wikis/Custom-Plugins) and use the existing plugins as guidelines as well.

**Comment by forthrin** [comment]
Haven't tried it yet. Wanted to ask before spending more time if it's done according to conventions and if it actually uses the correct classes, etc. Will submit it for the benefit of others when finished. So, comments?

**Comment by HoneyryderChuck** [comment]
guidelines-wise, seems ok. Don't think it works as is (do read the link I shared, and what each module is supposed to represent), so I'll let you iterate on that.
